Add basic sign-up page

LeonRabinovich opened the pull request:
Resolves: #88 

- Installed django-crispy-forms (used in the HTML pages)
- Sign-up page is available at 127.0.0.1:8000/**users/sign_up/**
- Sign-up uses Django built-in validations for username, password, password confirmation 
- If Sign-up is successful, page is redirected to the Sign-in page which is added in a different PR
 (over-wise the sign-up form is cleared)
- Implemented fail/success messages  

* The following test method will be added after sign-in page is merged:
**def test_post_valid_signup_with_client(self, client, valid_user_details):**

       `response = client.post('/users/sign_up/', data=valid_user_details)

        assert response.status_code == 302
        assert response.url == '/users/sign_in/' `

**sign-in, sign-up, sign-out demonstration video:**

https://user-images.githubusercontent.com/79373827/118169108-3b868180-b431-11eb-8261-ac304659dd6e.mp4

LeonRabinovich committed:
Adding django-crispy forms to the project

LeonRabinovich committed:
Adding basic sign-up page with tests

VeraTs commented on homepage/views.py:
```diff
@@ -53,3 +55,23 @@ def add_review(request):
+            return redirect("sign_in")
+
+        else:
+            messages.error(request, "Unsuccessful registration. Invalid information.")
```
Is there no way to determine which field is invalid, specifically? 
Or are there automated error messages for the django sign up form?

oriAdler replied:
Same like Vera, a more verbose error message will be user friendly

LeonRabinovich replied:
Regarding the comment on the test file - if i remove it the automatic tests on github will fail - so i prefer to keep it like that.

Regarding the messages - they are only partly implemented, so this will be improved in a separate PR.
 (added clarification on the PR comment )

VeraTs replied:
If you have an error message already, no matter how basic, that just says "oops, can't do that",
And it's implemented (as it is here)
then pass it on to render and five it the 3 lines in the html file.
Further implementation should include more specific error messages, if we want them,
but I believe that if you're calling upon it and giving it a value, it is ready for display.

LeonRabinovich replied:
@VeraTs 
I'm convinced - added the messages in the HTML file.

VeraTs reviewed (commented):
Overall okay, as said before, I like the addition of crisp forms
I think you should note in the PR that it's dependent on #99 and remove the comment from the test file, but that's just a suggestion.
I do want to know about the error message, as stated below, and also about displaying it in the form, as I stated in PR #99

oriAdler commented on homepage/views.py:
```diff
@@ -53,3 +55,23 @@ def add_review(request):
+            appUser.save()
+
+            login(request, appUser.user)
+            messages.success(request, "Registration successful.")
```
Where is that message shown? if the user is redirected to the **sign_in** page

LeonRabinovich replied:
Error messages are only partly implemented and will be fully implemented on a separate PR with other features
I've added a comment in the PR description for more clarity.

oriAdler reviewed (commented):
Good work.
Can you add a screenshot of the new page?
If this PR is dependent on #99, I would suggest mentioning it and remove the commented tests.

LeonRabinovich commented:
> 
> 
> Good work.
> Can you add a screenshot of the new page?
> If this PR is dependent on #99, I would suggest mentioning it and remove the commented tests.

I prefer to keep the single commented test - over-wise this PR cannot be merged on its own because it will not pass the tests.
its not dependent on the other PR other than this test so i think its better this way.

oriAdler commented:
@LeonRabinovich I don't think it's best practice to merge commented code. 
So the suggestion is to remove the commented tests (you can save them locally).
BTW the ```resolve conversation``` is cool, but maybe you should wait for some indication from us (thumbs up) before, so you know it's really resolved.

LeonRabinovich commented:
> 
> 
> @LeonRabinovich I don't think it's best practice to merge commented code.
> So the suggestion is to remove the commented tests (you can save them locally).
> BTW the `resolve conversation` is cool, but maybe you should wait for some indication from us (thumbs up) before, so you know it's really resolved.



> 
> 
> @LeonRabinovich I don't think it's best practice to merge commented code.
> So the suggestion is to remove the commented tests (you can save them locally).
> BTW the `resolve conversation` is cool, but maybe you should wait for some indication from us (thumbs up) before, so you know it's really resolved.

Removed the commented test.

> 
> 
> @LeonRabinovich I don't think it's best practice to merge commented code.
> So the suggestion is to remove the commented tests (you can save them locally).
> BTW the `resolve conversation` is cool, but maybe you should wait for some indication from us (thumbs up) before, so you know it's really resolved.

Removed the commented test.

LeonRabinovich commented:
> 
> 
> @LeonRabinovich I don't think it's best practice to merge commented code.
> So the suggestion is to remove the commented tests (you can save them locally).
> BTW the `resolve conversation` is cool, but maybe you should wait for some indication from us (thumbs up) before, so you know it's really resolved.

Removed the commented test.

VeraTs commented:
@LeonRabinovich Please un-resolve the thread of my initial comment, and continue the conversation there, since I'm not satisfied that what you decided is the best course of action.

LeonRabinovich commented:
> 
> 
> @LeonRabinovich Please un-resolve the thread of my initial comment, and continue the conversation there, since I'm not satisfied that what you decided is the best course of action.

Unsolved, will reply there

LeonRabinovich committed:
Removing blank line from methods

LeonRabinovich committed:
Adding messages to the HTML page

LeonRabinovich committed:
Merge branch 'main' into signUp, fix conflicts

LeonRabinovich committed:
Fix mistake in app_layout when solving conflict

Yarboa reviewed (approved):
LGTM

VeraTs commented on homepage/templates/homepage/app_layout.html:
```diff
@@ -48,6 +48,24 @@
+      {% endfor %}
+  </ul>
+  {% endif %}
+
```
lines 52 - 66 :
as in PR #99

LeonRabinovich replied:
@VeraTs 
see my reply in PR #99

VeraTs reviewed (commented):
I'm happy you agreed to the error comment aspect, my only comment here is the same as in PR #99

LeonRabinovich committed:
Merge branch 'main' into signUp, fix conflicts

LeonRabinovich committed:
Fixing the style implementation of messages in app_layout

kobihk commented on homepage/templates/homepage/app_layout.html:
```diff
@@ -32,8 +37,8 @@
-        <a class="nav-link disabled" href="#">Sign In</a>
-      </li>
+        <a class="nav-link" href="/users/sign_in">Sign In</a>
+      </li>        
```
looks like you have extra spaces in the end of the line please remove them
```suggestion
      </li> 

```

LeonRabinovich replied:
removed

kobihk commented on homepage/templates/homepage/app_layout.html:
```diff
@@ -48,6 +53,16 @@
+  {% if messages %}
+  <ul class="messages">
+      {% for message in messages %}
+      <li{% if message.tags %} class="{{ message.tags }}"{% endif %}>{{ message }}</li>
```
please check that in case of message.tags = None we don't get empty list item like:

-

LeonRabinovich replied:
> please check that in case of message.tags = None we don't get empty list item like:

@kobihk 
Can you please elaborate regarding this ? I'm not sure i understand.

From what i see in the django message documentation, each message must have a tag, and the tag represents the message level and also some additional information about the message if needed.
I dont think if message.tags == None then we will get an empty list displayed (i could not reproduce tags==None)
https://docs.djangoproject.com/en/3.2/ref/contrib/messages/#message-tags

I currently use the implementation from the django documentation -
https://docs.djangoproject.com/en/3.2/ref/contrib/messages/#displaying-messages


Also - how do i need to change the code above to check the case of message.tag == None?
right now i see there is a check `{% if message.tags %}`

kobihk replied:
opss...
I missed the {{ message }} at the end of the line

kobihk commented on homepage/views.py:
```diff
@@ -77,3 +79,22 @@ def add_review_search(request):
+        form = SignUpForm(request.POST)
+
+        if form.is_valid():
+            appUser = AppUser()
```
please follow the python convention
```suggestion
            app_user = AppUser()
```

LeonRabinovich replied:
changed

LeonRabinovich committed:
Removing space, fixing python names

LeonRabinovich committed:
Merge branch 'main' into signUp, fix conflicts

kobihk commented:
@LeonRabinovich please resolve the conflicts

LeonRabinovich committed:
Merge branch 'main' into signUp, fix conflicts

LeonRabinovich committed:
Fixing some test names and adding a test that depends on sign-in

LeonRabinovich commented:
> @LeonRabinovich please solve the conflicts

@kobihk 
Fixed
